google-apis-core/lib/google/apis/core/storage_upload.rb

```diff
@@ -0,0 +1,181 @@
+          end
+        end
+
+        # Close IO stream when command done. Only closes the stream if it was opened by the command.
+        def release!
+          upload_io.close if @close_io_on_finish
+        end
+
+        # Execute the command, retrying as necessary
+        #
+        # @param [HTTPClient] client
+        #   HTTP client
+        # @yield [result, err] Result or error if block supplied
+        # @return [Object]
+        # @raise [Google::Apis::ServerError] An error occurred on the server and the request can be retried
+        # @raise [Google::Apis::ClientError] The request is invalid and should not be retried without modification
+        # @raise [Google::Apis::AuthorizationError] Authorization is required
+        def execute(client)
+          prepare!
+          opencensus_begin_span
+
+          do_retry :initiate_resumable_upload, client
+          while @upload_incomplete
+            res = do_retry :send_upload_command, client
+          end
+          res
+        ensure
+          opencensus_end_span
+          @http_res = nil
+          release!
+        end
+
+        def initiate_resumable_upload(client)
+          logger.debug { sprintf('Intiating resumable upload command to %s', url) }
+
+          request_header = header.dup
+          apply_request_options(request_header)
+
+          request_query = query.dup
+          request_query['uploadType'] = RESUMABLE
+
+          request_header[CONTENT_LENGTH_HEADER] = upload_io.size.to_s
+          request_header[CONTENT_TYPE_HEADER] = JSON_CONTENT_TYPE
+          request_header[UPLOAD_CONTENT_TYPE_HEADER] = upload_content_type unless upload_content_type.nil?
+
+          response = client.post(url.to_s, query: request_query,
+                         body: body,
+                         header: request_header,
+                         follow_redirect: true)
+          result = process_response(response.status_code, response.header, response.body)
+          success(result)
+        rescue => e
+          error(e, rethrow: true)
+        end
+
+        # Send the actual content
+        #
+        # @param [HTTPClient] client
+        #   HTTP client
+        # @return [HTTP::Message]
+        # @raise [Google::Apis::ServerError] Unable to send the request
+        def send_upload_command(client)
+          logger.debug { sprintf('Sending upload command to %s', @upload_url) }
+
+          remaining_content_size = upload_io.size - @offset
+          current_chunk_size = remaining_content_size < CHUNK_SIZE ? remaining_content_size : CHUNK_SIZE
+
+          request_header = header.dup
+          request_header[CONTENT_RANGE_HEADER] = sprintf('bytes %d-%d/%d', @offset, @offset+current_chunk_size-1, upload_io.size)
+          request_header[CONTENT_LENGTH_HEADER] = current_chunk_size
+          body = upload_io.read(current_chunk_size)
+
+          response = client.put(@upload_url, body: body, header: request_header, follow_redirect: true)
```

Comment:
Small nit here, but I'd name the `body` local variable something like `chunk_body` instead, to distinguish it from `self.body` which is a method on the base class (and is treated as such in line 120). This will protect you from a bug if you ever move these lines around and forget that the name is overloaded.

Reply:
Makes sense. Thanks for pointing it out.